packages/react-hooks/src/useControllableValue.ts

```diff
@@ -0,0 +1,37 @@
+import * as React from 'react';
+
+/**
+ * Hook to manage a value that could be either controlled or uncontrolled, such as a checked state or
+ * text box string.
+ * @param controlledValue- The controlled value passed in the props. This value will always be used if provided, and the
+ * internal state will be updated to reflect it.
+ * @param defaultUncontrolledValue- Initial value for the internal state in the uncontrolled case.
+ * @see https://reactjs.org/docs/uncontrolled-components.html
+ */
+export function useControllableValue<TValue, TOnChangeArgs extends unknown[]>(
+  controlledValue: TValue | undefined,
+  defaultUncontrolledValue: TValue | undefined,
+  onChange?: (newValue: TValue, ...args: TOnChangeArgs) => void,
+) {
+  const [value, setValue] = React.useState<TValue | undefined>(
+    controlledValue !== undefined ? controlledValue : defaultUncontrolledValue,
+  );
+
+  if (controlledValue !== undefined && controlledValue !== value) {
+    setValue(controlledValue);
```

Comment:
is this line needed? If `controlledValue` is provided, that's the state, period. Setting the value here would potentially cause a re-render unnecessarily i'd think.

Reply:
@dzearing It's copied from the implementation in `Checkbox.Base.tsx`. It also ensures that if the controlled value "goes missing" on a future render, the state remains consistent.

Reply:
I believe this ends up penalizing every controlled render which wouldn't be shippable. I'd recommend removing it, and maybe even adding a test to validate that a component's updating of a controlled value equates to exactly 1 render.

If the value ever goes missing, I would suggest a debug-only warning if you want to protect against this scenario.